crates/transaction-pool/src/pool/pending.rs

```diff
@@ -325,10 +350,58 @@ impl<T: TransactionOrdering> PendingPool<T> {
         id
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.all.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Truncates the pool to the given limit.
+    ///
+    /// Removes transactions from the pending pool
+    /// The algorithm tries to reduce transaction counts by an approximately
+    /// equal number for all for accounts with many pending transactions.
+    pub(crate) fn truncate_pool(
+        &mut self,
+        limit: SubPoolLimit,
+    ) -> Vec<Arc<ValidPoolTransaction<T::Transaction>>> {
+        let mut removed = Vec::new();
+
+        // penalize non-local txs if limit is still exceeded
+        while self.size() > limit.max_size || self.len() > limit.max_txs {
+            // flag to break out of the loop if all are local
+            let mut any_non_local = true;
+
+            // use descendants, so we pop a single tx per sender
+            for tx in self.independent_descendants.clone().iter() {
```

Comment:
hmm, I think the better approach here would be to not clone at all and instead collect the transactions we want to evict in a vec then remove